Two new tests for move only types

neil-ca-moore opened the pull request:
- Renamed test_helper.{hpp,cpp} to future_test_helper because the helpers were future specific
- Now test_helper.{hpp,cpp} are for global helpers
- Made a move_only_t type for testing move only behaviour (didn't use existing type like unique_ptr because I didn't want to make anything language version specific - that type wasn't move only until c++17)
- Added a couple of tests that move only types pass through futures and channels without any troubles

Not sure why future_constructed_minimal_fn_moveonly was disabled in the first place, I hope the test I wrote adequately replaces it.

Most of the changes are refactorings relating to test helpers, I hope this is OK.

FelixPetriconi commented on test/CMakeLists.txt:
```diff
@@ -1,6 +1,7 @@
 PROJECT(future_test)
 
 set(SOURCE
+  test_helper.cpp
```
Can we keep the alphabetical ordering?
As well for all #includes?

neil-ca-moore replied:
OK. main.cpp was out of order already though, do you want that changed?

FelixPetriconi commented on test/channel_tests.cpp:
```diff
@@ -198,6 +199,24 @@ BOOST_FIXTURE_TEST_SUITE(int_channel, channel_test_fixture_int_1)
         BOOST_REQUIRE_EQUAL(42, result);
     }
 
+    BOOST_AUTO_TEST_CASE(move_only_type_channel) {
```
There was already a set of move only tests in channel_functor_tests.cpp:151 ff,
so this is not really necessary, or?

neil-ca-moore replied:
No, I just missed it.

FelixPetriconi commented on test/future_test_helper.cpp:
```diff
@@ -0,0 +1,24 @@
+/*
```
Good, that you renamed it.

FelixPetriconi commented:
Regarding your comments:
I developed all the tests under Visual Studio 2015 and there was std::unique_ptr already move only.
The disabled test did not work at that time and I forgot to follow up on it. Thanks for finding it.

FelixPetriconi commented:
# ATTENTION # 
This PR goes to the master and not to develop

fosterbrereton commented:
I repointed the PR to develop.

neil-ca-moore commented:
Would you would prefer that I got rid of the move_only_t that I added and instead use unique_ptr? I added it mainly because I felt it made it more explicit what the test was doing.

sean-parent commented:
I added a list of things to test in issue #69. I'm in favor of move_only_t as a model type.

neil-ca-moore commented:
@sean-parent I'll do another PR with more tests if I'm able to work on that.

neil-ca-moore commented:
@FelixPetriconi I have addressed your comments.

neil-ca-moore commented:
This probably failed the test because I moved code from one file to another.

fosterbrereton commented:
@neil-ca-moore it still seems like the PR has some merge conflicts. Can you confirm the PR is as you intend? The conflicts seem pretty minor, but I'll defer to you to clean them out.

fosterbrereton commented on test/test_helper.hpp:
```diff
@@ -2,264 +2,28 @@
-
-        std::atomic_int _task_counter;
-
+    class move_only_t {
```
It would be great if this type could be moved to `stlab/test/model.hpp`. It would be useful for testing e.g., `task`.

neil-ca-moore replied:
I did that.

fosterbrereton replied:
Yes, you did! Thanks, sorry, I misread the diff.

neil-ca-moore replied:
No I only just did that 5 minutes ago.

fosterbrereton commented:
LGTM; Thanks for helping out!
